internal/sshtunneler/tunneler_test.go

```diff
@@ -0,0 +1,234 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package sshtunneler
+
+import (
+	"context"
+	"net"
+	"sync"
+	"time"
+
+	jc "github.com/juju/testing/checkers"
+	"github.com/lestrrat-go/jwx/v2/jwa"
+	gomock "go.uber.org/mock/gomock"
+	gc "gopkg.in/check.v1"
+
+	network "github.com/juju/juju/core/network"
+	"github.com/juju/juju/state"
+)
+
+type sshTunnelerSuite struct {
+	state      *MockState
+	controller *MockControllerInfo
+	dialer     *MockSSHDial
+	clock      *MockClock
+}
+
+var _ = gc.Suite(&sshTunnelerSuite{})
+
+func (s *sshTunnelerSuite) setupMocks(c *gc.C) *gomock.Controller {
+	ctrl := gomock.NewController(c)
+
+	s.state = NewMockState(ctrl)
+	s.controller = NewMockControllerInfo(ctrl)
+	s.dialer = NewMockSSHDial(ctrl)
+	s.clock = NewMockClock(ctrl)
+
+	return ctrl
+}
+
+func (s *sshTunnelerSuite) newTunnelTracker(c *gc.C) *TunnelTracker {
+	args := TunnelTrackerArgs{
+		State:          s.state,
+		ControllerInfo: s.controller,
+		Dialer:         s.dialer,
+		Clock:          s.clock,
+		SharedSecret:   []byte("test-secret"),
+		JWTAlg:         jwa.HS256,
+	}
+	tunnelTracker, err := NewTunnelTracker(args)
+	c.Assert(err, jc.ErrorIsNil)
+	return tunnelTracker
+}
+
+func (s *sshTunnelerSuite) TestTunneler(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	tunnelTracker := s.newTunnelTracker(c)
+
+	sshConnArgs := state.SSHConnRequestArg{}
+
+	now := time.Now()
+
+	s.controller.EXPECT().Addresses().Return([]network.SpaceAddress{
+		{MachineAddress: network.NewMachineAddress("1.2.3.4")},
+	})
+	s.state.EXPECT().InsertSSHConnRequest(gomock.Any()).DoAndReturn(
+		func(sra state.SSHConnRequestArg) error {
+			sshConnArgs = sra
+			return nil
+		},
+	)
+	s.dialer.EXPECT().Dial(gomock.Any(), gomock.Any(), gomock.Any()).Return(nil, nil)
+	s.clock.EXPECT().Now().AnyTimes().Return(now)
+
+	tunnelReqArgs := RequestArgs{
+		MachineID: "0",
+		ModelUUID: "model-uuid",
+	}
+
+	req, err := tunnelTracker.RequestTunnel(tunnelReqArgs)
+	c.Assert(err, jc.ErrorIsNil)
+
+	var tunnels []string
+	for uuid := range tunnelTracker.tracker {
+		tunnels = append(tunnels, uuid)
+	}
+	c.Assert(tunnels, gc.HasLen, 1)
+
+	tID, err := tunnelTracker.AuthenticateTunnel("reverse-tunnel", sshConnArgs.Password)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(tID, gc.Equals, tunnels[0])
+
+	wg := sync.WaitGroup{}
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		err := tunnelTracker.PushTunnel(context.Background(), tID, nil)
+		c.Check(err, jc.ErrorIsNil)
```

Comment:
hmm.. not sure we want to use c.Check in a goroutine.

Reply:
Check is okay in a go routine I believe, I think it's assert that shouldn't be used.